Update implicit-conversions.md (#2674)
scala/docs.scala-lang: _tour/implicit-conversions.md

An example is to compare two strings `"foo" < "bar"`. In this case, `String` has no member `<`, so the implicit conversion `Predef.augmentString("foo") < "bar"` is inserted. (`scala.Predef` is automatically imported into all Scala programs.)

### How are implicit conversions brought into scope? ###

{% tabs implicit-conversion-scope class=tabs-scala-version %}
{% tab 'Scala 2' %}

> In Scala 2, an implicit convesion is brought into scope by importing from `ImplicitConversions`.

Reviewer: ```suggestion
In Scala 2, an implicit conversion is brought into scope by importing from the scope it was defined in, (e.g. `Conversions` in this case).
```

```scala mdoc
case class Student(name: String) {
  def printName: Unit = println(name)
}
  
object Conversions {
  implicit def fromStringToStudent(name: String): Student = Student(name)
}

import ImplicitConversions._
object Usage:
  def main(args: Array[String]) = "Reginald".printName
```
{% endtab %}
{% tab 'Scala 3' %}

> In Scala 3, an implicit conversion must be explicitly imported in order to be applicable.

Reviewer: this isn't true, you can also have a given import:
```scala
import Conversions.given
import Conversions.{given Conversion[String, Student]}
```

Reviewer: or if the implicit conversion is in the companion object of the argument type, (e.g. String in this case) then no import is necessary

Author: Thank you for all the suggestions! I've learned a lot! I've also improved the documentation.

Author: Just asking: There's no way to add a companion object to a Java class like `java.lang.String`, is it? I tried that but either the compiler runs into trouble or it does not find the conversion.

Reviewer: companion objects must be defined in the same source file :)

{% tab 'Scala 2' %}
In Scala 2, an implicit conversion is brought into scope by importing from the object that defined it, (e.g. `Conversions` in this case).

```scala mdoc
case class Student(name: String) {
  def printName: Unit = println(name)
}
  
object Conversions {
  implicit def fromStringToStudent(name: String): Student = Student(name)
}

import Conversions._
object Usage:
  def main(args: Array[String]) = "Reginald".printName
```
{% endtab %}
{% tab 'Scala 3' %}
In Scala 3, an implicit conversion is brought into scope by either importing `given` or the named conversion from the object that defined it, (e.g. `Conversions` in this case). 

Note that as of Scala 3, implicit conversions cannot be brought into scope anymore by means of a wildcard import (`*`). 

Given the example:

```scala mdoc
case class Student(name: String):
  def printName: Unit = println(name)
object Student:
  given Conversion[Student, Int] = _.name.length
  
object Conversions:
  given fromStringToStudent: Conversion[String, Student] = Student(_)
```

The following imports would bring the `Conversion[String, Student]` into scope:
  - `import Conversions.given`

>   - `import Conversions.{given[String, Student]}`

Reviewer: ```suggestion
  - `import Conversions.{given Conversion[String, Student]}`
```

  - `import Conversions.fromStringToStudent`

> If the implicit conversion is in the companion object of the argument type, (e.g. `Student` in this case), then no import is necessary.

Reviewer: this also applies to Scala 2

Author: Thanks, I added it to the Scala 2 example as well.

An example is to compare two strings `"foo" < "bar"`. In this case, `String` has no member `<`, so the implicit conversion `Predef.augmentString("foo") < "bar"` is inserted. (`scala.Predef` is automatically imported into all Scala programs.)

### How are implicit conversions brought into scope? ###

{% tabs implicit-conversion-scope class=tabs-scala-version %}
{% tab 'Scala 2' %}
In Scala 2, an implicit conversion is brought into scope by importing from the object that defined it, (e.g. `Conversions` in this case). If the implicit conversion is in the companion object of the argument type, (e.g. `Student` in this case), then no import is necessary.

```scala mdoc

> case class Student(name: String)

Reviewer: ```suggestion
import scala.language.implicitConversions

case class Student(name: String)
```

  def main(args: Array[String]) = {
    val reginald: Student = "Reginald" // applies the conversion Conversions.fromStringToStudent
    println(reginald + 2)              // applies the conversion Student.fromStudentToInt
  }
}
```
{% endtab %}
{% tab 'Scala 3' %}
In Scala 3, an implicit conversion is brought into scope by either importing `given` or the named conversion from the object that defined it, (e.g. `Conversions` in this case). 

Note that as of Scala 3, implicit conversions cannot be brought into scope anymore by means of a wildcard import (`*`). 

Given the example:

```scala
case class Student(name: String):
  def printName: Unit = println(name)
object Student:
  given Conversion[Student, Int] = _.name.length
  
object Conversions:
  given fromStringToStudent: Conversion[String, Student] = Student(_)
```

The following imports would bring the `Conversion[String, Student]` into scope:
  - `import Conversions.given`
  - `import Conversions.{given Conversion[String, Student]}`
  - `import Conversions.fromStringToStudent`

If the implicit conversion is in the companion object of the argument type, (e.g. `Student` in this case), then no import is necessary.

> ```scala mdoc

Reviewer: ```suggestion
```scala
```
